neonvm/controllers/virtualmachine_controller.go

```diff
@@ -744,55 +744,109 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
 		}
 
 	case vmv1.VmSucceeded, vmv1.VmFailed:
-		switch virtualmachine.Spec.RestartPolicy {
-		case vmv1.RestartPolicyAlways:
-			log.Info("Restarting VM runner pod", "VM.Phase", virtualmachine.Status.Phase, "RestartPolicy", virtualmachine.Spec.RestartPolicy)
-			// get runner to delete
-			vmRunner := &corev1.Pod{}
-			err := r.Get(ctx, types.NamespacedName{Name: virtualmachine.Status.PodName, Namespace: virtualmachine.Namespace}, vmRunner)
-			if err == nil {
-				// delete current runner
-				if err := r.deleteRunnerPodIfEnabled(ctx, virtualmachine, vmRunner); err != nil {
-					return err
-				}
-			} else if !apierrors.IsNotFound(err) {
+		// Always delete runner pod. Otherwise, we could end up with one container succeeded/failed
+		// but the other one still running (meaning that the pod still ends up Running).
+		vmRunner := &corev1.Pod{}
+		err := r.Get(ctx, types.NamespacedName{Name: virtualmachine.Status.PodName, Namespace: virtualmachine.Namespace}, vmRunner)
```

Comment:
When dealing with state machines like this, it's easier to track down what's going on where each case is handled separately. Now it's kinda hard to tell what we do in case of `vmv1.RestartPolicyAlways` and `vmv1.RestartPolicyAlways`. I suggest handling each case separately as before.

Reply:
Fair enough, reintroduced that explicit handling but kept the deduplication for the restarting logic (5c19ad6601ccedc2e2d1e32418b816dd1e005acd). wdyt?

Reply:
Looks good